File: py/662.maximum-width-of-binary-tree.py
#
# @lc app=leetcode id=662 lang=python3
#
# [662] Maximum Width of Binary Tree
#
from typing import *
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        node_pos_queue = deque()
        node_pos_queue.append((root, 0))
        res = 0
        while(len(node_pos_queue)):
            q_len = len(node_pos_queue)
            res = max(res, node_pos_queue[-1][1]-node_pos_queue[0][1])
            shift = node_pos_queue[0][1]
            while q_len:
                q_len -= 1
                elem = node_pos_queue.popleft()
                if elem[0].left:
                    node_pos_queue.append((elem[0].left, (elem[1]-shift)*2))
                if elem[0].right:
                    node_pos_queue.append((elem[0].right, (elem[1]-shift)*2+1))
        return res+1
# @lc code=end

File: cpp/662.maximum-width-of-binary-tree.cpp
/*
 * @lc app=leetcode id=662 lang=cpp
 *
 * [662] Maximum Width of Binary Tree
 */
#include <queue>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    int widthOfBinaryTree(TreeNode *root)
    {
        queue<pair<TreeNode *, long long>> node_pos_queue;
        node_pos_queue.push({root, 0});
        long long res = 0;
        while (!node_pos_queue.empty())
        {
            int q_len = node_pos_queue.size();
            res = max(res, node_pos_queue.back().second - node_pos_queue.front().second);
            int shift = node_pos_queue.front().second;
            while (q_len--)
            {
                auto elem = node_pos_queue.front();
                node_pos_queue.pop();
                if (elem.first->left)
                {
                    node_pos_queue.push({elem.first->left, (elem.second - shift) * 2});
                }
                if (elem.first->right)
                {
                    node_pos_queue.push({elem.first->right, (elem.second - shift) * 2 + 1});
                }
            }
        }
        return res + 1;
    }
};
// @lc code=end
